core/src/impl/Kokkos_Default_GraphNodeKernel.hpp

```diff
@@ -0,0 +1,124 @@
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_KOKKOS_HOST_GRAPHNODEKERNEL_HPP
+#define KOKKOS_KOKKOS_HOST_GRAPHNODEKERNEL_HPP
+
+#include <Kokkos_Macros.hpp>
+
+#include <impl/Kokkos_Default_Graph_fwd.hpp>
+
+#include <Kokkos_Graph.hpp>
+#include <Kokkos_Parallel.hpp>
+#include <Kokkos_Parallel_Reduce.hpp>
+
+namespace Kokkos {
+namespace Impl {
+
+//==============================================================================
+// <editor-fold desc="GraphNodeKernelImpl"> {{{1
+
+template <class ExecutionSpace>
+struct GraphNodeKernelDefaultImpl {
+  // TODO @graphs decide if this should use vtable or intrusive erasure via
+  //      function pointers like in the rest of the graph interface
+  virtual void execute_kernel() const = 0;
+};
+
+// TODO Indicate that this kernel specialization is only for the Host somehow?
+template <class ExecutionSpace, class PolicyType, class Functor,
+          class PatternTag, class... Args>
+class GraphNodeKernelImpl
+    : public PatternImplSpecializationFromTag<PatternTag, Functor, PolicyType,
+                                              Args..., ExecutionSpace>::type,
+      public GraphNodeKernelDefaultImpl<ExecutionSpace> {
+ public:
+  using base_t =
+      typename PatternImplSpecializationFromTag<PatternTag, Functor, PolicyType,
+                                                Args..., ExecutionSpace>::type;
+  using execute_kernel_vtable_base_t =
+      GraphNodeKernelDefaultImpl<ExecutionSpace>;
+  // We have to use this name here because that's how it was done way back when
+  // then implementations of Impl::Parallel*<> were written
+  using Policy       = PolicyType;
+  using graph_kernel = GraphNodeKernelImpl;
+
+  // TODO @graph kernel name info propagation
+  template <class PolicyDeduced, class... ArgsDeduced>
+  GraphNodeKernelImpl(std::string, ExecutionSpace const&, Functor arg_functor,
```

Comment:
Take the `string` by const reference here, because we aren't using it.

Reply:
Resolved in 8e37a190